#! /bin/sh -e

# All lines beginning with `# DP:' are a description of the patch.
# DP: Description: Fix compilation against new binutils.
# DP: Related bugs: 
# DP: Dpatch author: Jeff Bailey
# DP: Patch author: HJ Lu
# DP: Upstream status: Pending
# DP: Status Details:
# DP: Date: 2005-16-11

PATCHLEVEL=1

if [ $# -ne 2 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch) patch -d "$2" -f --no-backup-if-mismatch -p$PATCHLEVEL < $0;;
    -unpatch) patch -d "$2" -f --no-backup-if-mismatch -R -p$PATCHLEVEL < $0;;
    *)
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
esac
exit 0

# append the patch here and adjust the -p? flag in the patch calls.

--- libc/sysdeps/wordsize-32/Makefile.as     2002-09-18 23:47:56.000000000 -0700
+++ libc/sysdeps/wordsize-32/Makefile        2005-11-07 15:13:38.000000000 -0800
@@ -2,5 +2,6 @@ ifeq ($(subdir),csu)
 ifeq (yes,$(build-shared))
 sysdep_routines += divdi3
 shared-only-routines += divdi3
+CPPFLAGS-divdi3.c = -Din_divdi3_c
 endif
 endif
--- libc/sysdeps/wordsize-32/symbol-hacks.h.as       2004-03-08 12:59:34.000000000 -0800
+++ libc/sysdeps/wordsize-32/symbol-hacks.h  2005-11-07 14:57:49.000000000 -0800
@@ -22,7 +22,7 @@
    therefore we get PLTs.  Unnecessarily so.  Changing gcc is a big
    task which might not be worth it so we play tricks with the
    assembler.  */
-#if !defined __ASSEMBLER__ && !defined NOT_IN_libc && defined SHARED
+#if !defined __ASSEMBLER__ && !defined in_divdi3_c && !defined NOT_IN_libc && defined SHARED
 asm ("__divdi3 = __divdi3_internal");
 asm ("__udivdi3 = __udivdi3_internal");
 asm ("__moddi3 = __moddi3_internal");

===================================================================
--- libc/sysdeps/powerpc/powerpc32/fpu/s_lround.S.~1.4.~        2005-06-20 11:05:09.000000000 +0200
+++ libc/sysdeps/powerpc/powerpc32/fpu/s_lround.S       2005-11-18 19:24:09.000000000 +0100
@@ -79,7 +79,6 @@ ENTRY (__lround)
        b       .L9
        END (__lround)
 
-strong_alias (__lround, __lround)
 weak_alias (__lround, lround)
 
 strong_alias (__lround, __lroundf)
--- libc/sysdeps/unix/sysv/linux/powerpc/powerpc32/socket.S.~1.4.~      2005-06-20 11:05:20.000000000 +0200
+++ libc/sysdeps/unix/sysv/linux/powerpc/powerpc32/socket.S     2005-11-18 18:25:51.000000000 +0100
@@ -44,7 +44,11 @@
 #define stackblock 20
 
 #ifndef __socket
-#define __socket P(__,socket)
+# ifndef NO_WEAK_ALIAS
+#  define __socket P(__,socket)
+# else
+#  define __socket socket
+# endif
 #endif
 
        .text
@@ -114,4 +118,6 @@ ENTRY(__socket)
 
 PSEUDO_END (__socket)
 
+#ifndef NO_WEAK_ALIAS
 weak_alias (__socket, socket)
+#endif
--- libc/sysdeps/unix/sysv/linux/powerpc/powerpc64/socket.S.~1.5.~      2005-01-10 10:41:04.000000000 +0100
+++ libc/sysdeps/unix/sysv/linux/powerpc/powerpc64/socket.S     2005-11-18 19:10:23.000000000 +0100
@@ -41,12 +41,12 @@
 
 #define stackblock 80 /* offset to socket parm area.  */
 
-#ifndef socket
-/* If this is just socket.S leave it alone! */
-#else
 #ifndef __socket
-#define __socket P(__,socket)
-#endif
+# ifndef NO_WEAK_ALIAS
+#  define __socket P(__,socket)
+# else
+#  define __socket socket
+# endif
 #endif
 
        .text
@@ -120,4 +120,6 @@ ENTRY(__socket)
        cfi_endproc
 PSEUDO_END (__socket)
 
+#ifndef NO_WEAK_ALIAS
 weak_alias (__socket, socket)
+#endif
--- libc/sysdeps/unix/sysv/linux/sparc/sparc32/socket.S	2005-11-24 07:30:53.000000000 -0500
+++ libc/sysdeps/unix/sysv/linux/sparc/sparc32/socket.S	2005-11-24 07:31:14.000000000 -0500
@@ -40,7 +40,11 @@
    The .S files for the other calls just #define socket and #include this.  */
 
 #ifndef __socket
-#define __socket P(__,socket)
+# ifndef NO_WEAK_ALIAS
+#  define __socket P(__,socket)
+# else
+#  define __socket socket
+# endif
 #endif
 
 .globl __socket
@@ -105,4 +109,6 @@
 
 END (__socket)
 
+#ifndef NO_WEAK_ALIAS
 weak_alias (__socket, socket)
+#endif
--- libc/sysdeps/unix/sysv/linux/sparc/sparc64/socket.S	2005-11-24 07:34:58.000000000 -0500
+++ libc/sysdeps/unix/sysv/linux/sparc/sparc64/socket.S	2005-11-24 07:35:17.000000000 -0500
@@ -40,7 +40,11 @@
    The .S files for the other calls just #define socket and #include this.  */
 
 #ifndef __socket
-#define __socket P(__,socket)
+# ifndef NO_WEAK_ALIAS
+#  define __socket P(__,socket)
+# else
+#  define __socket socket
+# endif
 #endif
 
 .globl __socket
@@ -107,4 +111,6 @@
 
 END (__socket)
 
+#ifndef NO_WEAK_ALIAS
 weak_alias (__socket, socket)
+#endif
